### ISSUE TITLE
Rule Change: no-shadow does not report function declaration shadowing const

### ISSUE THREAD

**Issue by lukpsaxo** [eslint/eslint#20425]
### What rule do you want to change?

no-shadow

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

[playground](https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gd3JhcCh5KSB7XG4gIHJldHVybiAoKSA9PiB7XG4gICAgY29uc29sZS5sb2coJ3dyYXBwZWQnKTtcbiAgICB5KCk7XG4gIH07XG59XG5sZXQgeCA9IGZhbHNlO1xuZXhwb3J0IGNvbnN0IGEgPSB3cmFwKGZ1bmN0aW9uIGEoKSB7XG4gIGlmICgheCkge1xuICAgIHggPSB0cnVlO1xuICAgIGEoKTtcbiAgfVxufSk7XG5cbiIsIm9wdGlvbnMiOnsicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOlsiZXJyb3IiXSwiZm9yLWRpcmVjdGlvbiI6WyJlcnJvciJdLCJnZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOlsiZXJyb3IiXSwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOlsiZXJyb3IiXSwibm8tY2xhc3MtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbXBhcmUtbmVnLXplcm8iOlsiZXJyb3IiXSwibm8tY29uZC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tY29uc3QtYXNzaWduIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWJpbmFyeS1leHByZXNzaW9uIjpbImVycm9yIl0sIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6WyJlcnJvciJdLCJuby1jb250cm9sLXJlZ2V4IjpbImVycm9yIl0sIm5vLWRlYnVnZ2VyIjpbImVycm9yIl0sIm5vLWRlbGV0ZS12YXIiOlsiZXJyb3IiXSwibm8tZHVwZS1hcmdzIjpbImVycm9yIl0sIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6WyJlcnJvciJdLCJuby1kdXBlLWVsc2UtaWYiOlsiZXJyb3IiXSwibm8tZHVwZS1rZXlzIjpbImVycm9yIl0sIm5vLWR1cGxpY2F0ZS1jYXNlIjpbImVycm9yIl0sIm5vLWVtcHR5IjpbImVycm9yIl0sIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1lbXB0eS1wYXR0ZXJuIjpbImVycm9yIl0sIm5vLWVtcHR5LXN0YXRpYy1ibG9jayI6WyJlcnJvciJdLCJuby1leC1hc3NpZ24iOlsiZXJyb3IiXSwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjpbImVycm9yIl0sIm5vLWZhbGx0aHJvdWdoIjpbImVycm9yIl0sIm5vLWZ1bmMtYXNzaWduIjpbImVycm9yIl0sIm5vLWdsb2JhbC1hc3NpZ24iOlsiZXJyb3IiXSwibm8taW1wb3J0LWFzc2lnbiI6WyJlcnJvciJdLCJuby1pbnZhbGlkLXJlZ2V4cCI6WyJlcnJvciJdLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6WyJlcnJvciJdLCJuby1sb3NzLW9mLXByZWNpc2lvbiI6WyJlcnJvciJdLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6WyJlcnJvciJdLCJuby1uZXctbmF0aXZlLW5vbmNvbnN0cnVjdG9yIjpbImVycm9yIl0sIm5vLW5vbm9jdGFsLWRlY2ltYWwtZXNjYXBlIjpbImVycm9yIl0sIm5vLW9iai1jYWxscyI6WyJlcnJvciJdLCJuby1vY3RhbCI6WyJlcnJvciJdLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOlsiZXJyb3IiXSwibm8tcmVkZWNsYXJlIjpbImVycm9yIl0sIm5vLXJlZ2V4LXNwYWNlcyI6WyJlcnJvciJdLCJuby1zZWxmLWFzc2lnbiI6WyJlcnJvciJdLCJuby1zZXR0ZXItcmV0dXJuIjpbImVycm9yIl0sIm5vLXNoYWRvdy1yZXN0cmljdGVkLW5hbWVzIjpbImVycm9yIl0sIm5vLXNwYXJzZS1hcnJheXMiOlsiZXJyb3IiXSwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOlsiZXJyb3IiXSwibm8tdW5kZWYiOlsiZXJyb3IiXSwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOlsiZXJyb3IiXSwibm8tdW5yZWFjaGFibGUiOlsiZXJyb3IiXSwibm8tdW5zYWZlLWZpbmFsbHkiOlsiZXJyb3IiXSwibm8tdW5zYWZlLW5lZ2F0aW9uIjpbImVycm9yIl0sIm5vLXVuc2FmZS1vcHRpb25hbC1jaGFpbmluZyI6WyJlcnJvciJdLCJuby11bnVzZWQtbGFiZWxzIjpbImVycm9yIl0sIm5vLXVudXNlZC1wcml2YXRlLWNsYXNzLW1lbWJlcnMiOlsiZXJyb3IiXSwibm8tdW51c2VkLXZhcnMiOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1iYWNrcmVmZXJlbmNlIjpbImVycm9yIl0sIm5vLXVzZWxlc3MtY2F0Y2giOlsiZXJyb3IiXSwibm8tdXNlbGVzcy1lc2NhcGUiOlsiZXJyb3IiXSwibm8td2l0aCI6WyJlcnJvciJdLCJyZXF1aXJlLXlpZWxkIjpbImVycm9yIl0sInVzZS1pc25hbiI6WyJlcnJvciJdLCJ2YWxpZC10eXBlb2YiOlsiZXJyb3IiXSwibm8tc2hhZG93IjpbImVycm9yIix7ImhvaXN0IjoiYWxsIn1dfSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=)

```js
function wrap(y) {
  return () => {
    console.log('wrapped');
    y();
  };
}
let x = false;
export const a = wrap(function a() {
  if (!x) {
    x = true;
    a();
  }
});
```

### What does the rule currently do for this code?

no errors

### What will the rule do after it's changed?

error on `a` - the function declaration is shadowing the const and importantly they are not equal to the same thing.

E.g. if I have

```
const a = function a() {}
```

I think it is ok to not give a no-shadow error.

but when it is:

```
const a = wrap(function a() {})
```

a !== a

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

This was the cause of a bug and I was surprised it wasn't caught by no-shadow.

I searched the issues for a long time, but its hard to know the key-words to search for - apologies if this has been covered before.

**Comment by lukpsaxo** [comment]
I assume the issue is that: https://github.com/eslint/eslint/blob/main/lib/rules/no-shadow.js#L357

is too lenient on assuming that if it is a function expression in the initializer then they must be the same variable.

**Comment by mdjermanovic** [comment]
I think this is a bug (false negative).

This is indeed caused by the logic in `function isOnInitializer`, which aims to allow code such as `var a = function a() {}` since it's common for a function/class expression to have the same name as the variable to which it is assigned, and although these are technically two variables where one shadows the other, they're effectively the same as they are referring to the same function/class, thus not reported by this rule.

https://github.com/eslint/eslint/blob/f0cafe5f37e7765e9d8c2751b5f5d33107687009/lib/rules/no-shadow.js#L348-L375

However, I don't think this exception should apply to code such as `const a = wrap(function a() {})`.

The only test cases related to this logic (the only test cases that fail when `function isOnInitializer` is removed) are two where a function/class expression is directly assigned to the variable:

https://github.com/eslint/eslint/blob/f0cafe5f37e7765e9d8c2751b5f5d33107687009/tests/lib/rules/no-shadow.js#L30

https://github.com/eslint/eslint/blob/f0cafe5f37e7765e9d8c2751b5f5d33107687009/tests/lib/rules/no-shadow.js#L41-L44

@eslint/eslint-team thoughts about accepting this as a bug fix?

**Comment by DMartens** [comment]
I agree that this should be a bug fix for, as the likely intent was to only exclude `var a = function a() {}`. That exception makes sense as the variable refers to the same function regardless of whether it is from inside or outside the function but does not longer hold true when passing it to a function (in the example `wrap` or modify it in any way).

**Comment by mdjermanovic** [comment]
Okay, then marking this as accepted. I can work on this.
